ambry-store/src/main/java/com/github/ambry/store/BlobStore.java

```diff
@@ -772,6 +775,183 @@ public void delete(List<MessageInfo> infosToDelete) throws StoreException {
+        return storeBatchDeleteInfo;
+      }
+      synchronized (storeWriteLock) {
+        Offset currentIndexEndOffset = index.getCurrentEndOffset();
+        i = 0;
+        if (!currentIndexEndOffset.equals(indexEndOffsetBeforeCheck)) {
+          FileSpan fileSpan = new FileSpan(indexEndOffsetBeforeCheck, currentIndexEndOffset);
+          filteredInfos = new ArrayList<>();
+          for (MessageInfo info : infosToDelete) {
+            try {
+              IndexValue value =
+                  index.findKey(info.getStoreKey(), fileSpan, EnumSet.allOf(PersistentIndex.IndexEntryType.class));
+              if (value != null) {
+                // There are several possible cases that can exist here. Delete has to follow either PUT, TTL_UPDATE or UNDELETE.
+                // let EOBC be end offset before check, and [RECORD] means RECORD is optional
+                // 1. PUT [TTL_UPDATE DELETE UNDELETE] EOBC DELETE
+                // 2. PUT EOBC TTL_UPDATE
+                // 3. PUT EOBC DELETE UNDELETE: this is really extreme case
+                // From these cases, we can have value being DELETE, TTL_UPDATE AND UNDELETE, we have to deal with them accordingly.
+                if (value.getLifeVersion() == lifeVersions.get(i)) {
+                  if (value.isDelete()) {
+                    throw new StoreException(
+                        "Cannot delete id " + info.getStoreKey() + " since it is already deleted in the index.",
+                        StoreErrorCodes.ID_Deleted);
+                  }
+                  // value being ttl update is fine, we can just append DELETE to it.
+                } else {
+                  // For the extreme case, we log it out and throw an exception.
+                  logger.warn("Concurrent operation for id " + info.getStoreKey() + " in store " + dataDir
+                      + ". Newly added value " + value);
+                  throw new StoreException(
+                      "Cannot delete id " + info.getStoreKey() + " since there are concurrent operation while delete",
+                      StoreErrorCodes.Life_Version_Conflict);
+                }
+                indexValuesPriorToDelete.set(i, value);
+              }
+              i++;
+              filteredInfos.add(info);
+            } catch (StoreException e){
+              storeBatchDeleteInfo.addMessageErrorInfo(new MessageErrorInfo(info, e.getErrorCode()));
+              if (e.getErrorCode() == StoreErrorCodes.IOError) {
+                onError();
+              }
+            }
+          }
+          infosToDelete = filteredInfos;
+          if (infosToDelete.isEmpty()){
+            logger.trace("Batch Delete completely failed since all blobs failed");
+            return storeBatchDeleteInfo;
+          }
+        }
+        List<InputStream> inputStreams = new ArrayList<>(infosToDelete.size());
+        List<MessageInfo> updatedInfos = new ArrayList<>(infosToDelete.size());
+        i = 0;
```

Comment:
How are we using `i` ?
It was being incremented in previous blocks, then set to 0 here and its being incremented again in next blocks but we don't seem to use the value anywhere?